nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/avro/AvroReaderWithExplicitSchema.java

```diff
@@ -17,33 +17,61 @@
 
 package org.apache.nifi.avro;
 
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
 import java.io.EOFException;
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.SequenceInputStream;
 
 import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileStream;
 import org.apache.avro.generic.GenericDatumReader;
 import org.apache.avro.generic.GenericRecord;
 import org.apache.avro.io.BinaryDecoder;
 import org.apache.avro.io.DatumReader;
 import org.apache.avro.io.DecoderFactory;
-import org.apache.nifi.schema.access.SchemaNotFoundException;
+import org.apache.commons.io.input.TeeInputStream;
 import org.apache.nifi.serialization.MalformedRecordException;
 import org.apache.nifi.serialization.record.RecordSchema;
 
 public class AvroReaderWithExplicitSchema extends AvroRecordReader {
     private final InputStream in;
     private final RecordSchema recordSchema;
     private final DatumReader<GenericRecord> datumReader;
-    private final BinaryDecoder decoder;
+    private BinaryDecoder decoder;
     private GenericRecord genericRecord;
+    private DataFileStream<GenericRecord> dataFileStream;
 
-    public AvroReaderWithExplicitSchema(final InputStream in, final RecordSchema recordSchema, final Schema avroSchema) throws IOException, SchemaNotFoundException {
+    public AvroReaderWithExplicitSchema(final InputStream in, final RecordSchema recordSchema, final Schema avroSchema) throws IOException {
         this.in = in;
         this.recordSchema = recordSchema;
 
-        datumReader = new GenericDatumReader<GenericRecord>(avroSchema);
-        decoder = DecoderFactory.get().binaryDecoder(in, null);
+        datumReader = new GenericDatumReader<>(avroSchema);
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        TeeInputStream teeInputStream = new TeeInputStream(in, baos);
+        // Try to parse as a DataFileStream, if it works, glue the streams back together and delegate calls to the DataFileStream
```

Comment:
It might be of interest to reverse the try/catch block. I understand the intuition to determine first whether the schema is embedded since that's the order in which the bytes are encoded, but there is an embedded Avro schema strategy for a reason. 

From my understanding, a user who chooses to use AvroReaderWithExplicitSchema over AvroReaderWithEmbeddedSchema  is expecting to use an Explicit Schema to decode Avro data without an embedded schema and the exceptional case is that the data contains an embedded schema. The current implementation forces that user's Flow to propagate an exception each time, which may be expensive. Maybe it's negligible with sufficient batching of records, but just a nit-picking thought. The TeeInputStream construct is pretty neat though :)

Reply:
That's a fair point to reverse the try/catch, and I did try that. The problem was that no error occurs on the raw file until you actually read from it, which happens outside the constructor during a record read loop. Here the exception will occur when the AvroReader is configured correctly, but it only happens once per flow file so the expense is amortized over the size of the flow file as you pointed out. Also, at least in my experience users are more likely to be using an embedded schema and trying to also explicitly define it than they are to be using schemaless files. Doesn't make a lot of sense (and maybe it's just my perception/experience) but I wrote the Jira because of how often people ran into this issue.